sabnzbd/nzbstuff.py

```diff
@@ -1185,6 +1193,41 @@ def remove_extrapar(self, parfile):
 
     __re_quick_par2_check = re.compile('\.par2\W*', re.I)
 
+
+    @synchronized(IO_LOCK)
+    def prospecive_add(self):
+        """ Add par2 files to compensate for missing articles
+        """
+        # How many do we need?
+        bad = len(self.nzo_info.get('bad_art_log', []))
+        miss = len(self.nzo_info.get('missing_art_log', []))
+        killed = len(self.nzo_info.get('killed_art_log', []))
+        dups = len(self.nzo_info.get('dup_art_log', []))
+        total_need = bad + miss + killed + dups
+
+        # How many do we already have?
+        blocks_already = 0
+        for nzf in self.files:
+            # Only par2 files have a blocks attribute
+            if nzf.blocks:
+                blocks_already = blocks_already + int_conv(nzf.blocks)
+
+        # Need more?
+        if blocks_already < total_need:
+            # We have to find the right par-set
+            for parset in self.extrapars.keys():
+                if parset in nzf.filename and self.extrapars[parset]:
+                    extrapars_sorted = sorted(self.extrapars[parset], key=lambda x: x.blocks, reverse=True)
+                    # Loop untill we have enough
+                    while blocks_already < total_need and extrapars_sorted:
+                        # Add the first one
+                        new_nzf = extrapars_sorted.pop()
+                        self.add_parfile(new_nzf)
+                        self.extrapars[parset] = extrapars_sorted
+                        blocks_already = blocks_already + int_conv(new_nzf.blocks)
+                        logging.info('Prospectively added %s repair blocks to %s', new_nzf.blocks, self.final_name)
```

Comment:
Here!
Only when a prospective block is added, we reset all.

I think we could limit the resetting to only when it's a ```.par2``` file, but not sure enough to change it now before the final.